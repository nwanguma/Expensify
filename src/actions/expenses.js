import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  //this dispatch argument is provided by thunk
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});


// REMOVE_EXPENSE_ASYNC
export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// EDIT_EXPENSE_ASYNC
export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    })
  }
}

// SET_EXPENSE
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

// SET_EXPENSES_ASYNC
export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];

      snapshot.forEach(childSnapshot => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    });
  };
};
