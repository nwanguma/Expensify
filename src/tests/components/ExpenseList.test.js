import { ExpenseList } from '../../components/ExpenseList.js';
import { shallow } from 'enzyme';
import React from 'react';
import expenses from '../fixtures/expense.js';

test('Should render the expense list component without props', () => {
  const wrapper = shallow(<ExpenseList />);
  expect(wrapper).toMatchSnapshot();
})

test('Should render the expense list component with props', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
})