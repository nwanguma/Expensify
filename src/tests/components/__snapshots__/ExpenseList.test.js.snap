// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the expense list component with props 1`] = `
<div>
  <ExpenseListItem
    amount={6700}
    createdAt={"1970-02-01T00:00:00.000Z"}
    description="Drinks"
    id="0"
    key="0"
    note="Three casks of bordeux wine for the Mrs"
  />
  <ExpenseListItem
    amount={9000}
    createdAt={"1970-01-01T00:00:00.000Z"}
    description="Food"
    id="1"
    key="1"
    note="Food for the month of September"
  />
  <ExpenseListItem
    amount={84800}
    createdAt={"1969-12-01T00:00:00.000Z"}
    description="Vacation"
    id="2"
    key="2"
    note="A trip to South Sudan"
  />
</div>
`;

exports[`Should render the expense list component without props 1`] = `
<div>
  <p>
    No expenses to show
  </p>
</div>
`;
